gc.c

```diff
@@ -2954,6 +2954,42 @@ should_be_finalizable(VALUE obj)
     rb_check_frozen(obj);
 }
 
+struct should_not_capture_data {
+    VALUE obj;
+    VALUE set;
+    bool found;
+};
+
+static void
+should_not_capture_callback(const VALUE child, struct should_not_capture_data *data)
+{
+    if (child == data->obj)
+	data->found = true;
+
+    if (data->found)
+	return;
+
+    // Maintain a set of objects already searched, so that we don't follow a cycle
+    VALUE key = rb_obj_id(child);
+    if (rb_hash_has_key(data->set, key))
+	return;
+    rb_hash_aset(data->set, key, Qtrue);
```

Comment:
Should we just use the object itself as the hash key?

Reply:
When we use `rb_objspace_reachable_objects_from` we get internal objects - `RUBY_T_IMEMO` and I think specifically instruction sequences. These aren't hashable (don't respond to `hash`). I'd welcome a suggestion about alternative data structures to use instead of a `Hash` - could I use the `st` hash structure directly? Or some kind of identity hash?